### PR TITLE
refactor: Prepare for correcting validation of max tables and max columns per table

### PR DESCRIPTION
**Pull request by carols10cents**
This connects to https://github.com/influxdata/idpe/issues/18100 but doesn't actually fix it yet.

These refactorings makes the change easy, the next PR will make the easy change to fix that issue.

Bill Hader as Stefon on Saturday Night Live

This is IOx's hottest PR. It's got everything:

- Reduced test code duplication
- Extracting related code from `data_types` into its own module
- Adding a newtype so raw values can't be accidentally mixed up
- Defining application defaults as `Default` impls on the application types, rather than consts in the catalog (the database actually has default values that are different 😱 )
- Hobgoblins providing foolish consistency by having `max_tables` fields always appear before `max_columns_per_table` fields
- Putting all related validation logic into one spot
- Small commits that should be straightforward to review one at a time

### CODE REVIEW

File: compactor/src/components/namespaces_source/mock.rs [hunk 1]
```diff
@@ -182,17 +185,17 @@ mod tests {
                     ns: Namespace {
                         id,
                         name: "ns".to_string(),
-                        max_tables: 10,
-                        max_columns_per_table: 10,
+                        max_tables: MaxTables::new(10),
+                        max_columns_per_table: MaxColumnsPerTable::new(10),
```
Comment (domodwyer) [inline compactor/src/components/namespaces_source/mock.rs:189]: TYPES ❤️

File: data_types/src/service_limits.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+use thiserror::Error;
+
+/// Max tables allowed in a namespace.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, sqlx::Type)]
+#[sqlx(transparent)]
+pub struct MaxTables(i32);
+
+#[allow(missing_docs)]
+impl MaxTables {
+    pub const fn new(v: i32) -> Self {
```
Comment (domodwyer) [inline data_types/src/service_limits.rs:15]: How do you feel about making this unsigned? -1 tables is nonsense - an unfortunate possibility of the current code too.

Annoyingly we have to encode it as a signed integer for postgres, but perhaps we could do that with a serialisation wrapper that takes the unsigned application-type and converts it into the `i32` representation needed for postgres. Bit annoying, but at least we banish -1 tables.

We'd have to validate that the input value doesn't exceed the bounds of an i32 too :+1:

File: data_types/src/service_limits.rs [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+    NoValueSpecified,
+}
+
+impl TryFrom<Option<LimitUpdate>> for ServiceLimitUpdate {
+    type Error = ServiceLimitError;
+
+    fn try_from(limit_update: Option<LimitUpdate>) -> Result<Self, Self::Error> {
+        match limit_update {
+            Some(LimitUpdate::MaxTables(n)) => {
+                if n == 0 {
```
Comment (domodwyer) [inline data_types/src/service_limits.rs:125]: Do these need to be `n <= 0` to cover off -1 tables? :fire:

File: gossip_schema/src/lib.rs [hunk 4]
```diff
@@ -209,8 +209,8 @@ mod tests {
                     .unwrap()
                     .clone(),
             ),
-            max_columns_per_table: 1,
             max_tables: 2,
```
Comment (domodwyer) [inline gossip_schema/src/lib.rs:212]: SUCH CONSISTENCY

File: iox_catalog/src/metrics.rs [hunk 5]
```diff
@@ -139,8 +139,8 @@ decorate!(
         "namespace_get_by_id" = get_by_id(&mut self, id: NamespaceId, deleted: SoftDeletedRows) -> Result<Option<Namespace>>;
         "namespace_get_by_name" = get_by_name(&mut self, name: &str, deleted: SoftDeletedRows) -> Result<Option<Namespace>>;
         "namespace_soft_delete" = soft_delete(&mut self, name: &str) -> Result<()>;
-        "namespace_update_table_limit" = update_table_limit(&mut self, name: &str, new_max: i32) -> Result<Namespace>;
-        "namespace_update_column_limit" = update_column_limit(&mut self, name: &str, new_max: i32) -> Result<Namespace>;
+        "namespace_update_table_limit" = update_table_limit(&mut self, name: &str, new_max: MaxTables) -> Result<Namespace>;
+        "namespace_update_column_limit" = update_column_limit(&mut self, name: &str, new_max: MaxColumnsPerTable) -> Result<Namespace>;
```
Comment (domodwyer) [inline iox_catalog/src/metrics.rs:143]: Ah it's so much better....

File: router/src/namespace_resolver.rs [hunk 6]
```diff
@@ -8,6 +8,7 @@ use thiserror::Error;
 
 use crate::namespace_cache::NamespaceCache;
 
+#[cfg(test)]
```
Comment (domodwyer) [inline router/src/namespace_resolver.rs:11]: Nice :+1:

Always love making test code explicitly test code.

**Review by domodwyer — APPROVED** [review]
I LOVE IT

I suspect my comments are destined to be magicked away in the next PR anyway 🚀 

This is way way way way way way way bananas way way way nicer - thank you 🙏

File: data_types/src/service_limits.rs (hunk 2 above)
Comment (carols10cents) [reply data_types/src/service_limits.rs:15]: Yes, that is coming in the next PR. I decided to just do rearranging to get all the validation stuff in one place before actually adding it, but yes changing this to usize and converting to i32 in the postgres catalog impl is my plan :)

**Pull request merged**